clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -1388,6 +1393,37 @@ public void commitSync(Map<TopicPartition, OffsetAndMetadata> offsets, Duration
         }
     }
 
+    private void awaitPendingAsyncCommitsAndExecuteCommitCallbacks(Timer timer, boolean disableWakeup) {
+        if (lastPendingAsyncCommit == null) {
+            return;
+        }
+
+        try {
+            CompletableFuture<Void> futureToAwait;
+            if (!disableWakeup) {
+                // We don't want the wake-up trigger to complete our pending async commit future,
+                // so create new future here.
+                futureToAwait = new CompletableFuture<>();
+                lastPendingAsyncCommit.whenComplete((v, t) -> {
+                    if (t != null) {
+                        futureToAwait.completeExceptionally(t);
+                    } else {
+                        futureToAwait.complete(v);
+                    }
+                });
+                wakeupTrigger.setActiveTask(futureToAwait);
+            } else {
+                futureToAwait = lastPendingAsyncCommit;
+            }
+            ConsumerUtils.getResult(futureToAwait, timer);
+            lastPendingAsyncCommit = null;
+        } finally {
+            if (!disableWakeup) wakeupTrigger.clearTask();
+            timer.update();
+        }
```

Comment:
Do we want to clear out the `lastPendingAsyncCommit` in the `finally` block:

```suggestion
            ConsumerUtils.getResult(futureToAwait, timer);
        } finally {
            lastPendingAsyncCommit = null;
            if (!disableWakeup) wakeupTrigger.clearTask();
            timer.update();
        }
```

Reply:
I think always clearing it in `finally` would mean that `lastPendingAsyncCommit` is cleared even though we timed out or were woken up while waiting for it.

However, this brought up another issue - what happens when the async commit future completes exceptionally? We'd throw the exception here, but we shouldn't - the error will be handled inside the future. So basically here we want to wait for the async commit, not worrying about return value or exception. And then, the only cases why we fail here should be wake-up or time out, and in both cases, we should check again for the future to be completed the next time we trigger commit sync.